internal/collector/otel_collector_plugin.go

```diff
@@ -117,19 +121,190 @@ func (oc *Collector) Close(ctx context.Context) error {
 	slog.InfoContext(ctx, "Closing OTel Collector plugin")
 
 	if !oc.stopped {
+		slog.InfoContext(ctx, "Shutting down OTel Collector", "state", oc.service.GetState())
 		oc.stopped = true
 		oc.service.Shutdown()
+		oc.cancel()
+
+		settings := oc.config.Common
+		settings.MaxElapsedTime = maxTimeToWaitForShutdown
+		err := backoff.WaitUntil(ctx, oc.config.Common, func() error {
+			if oc.service.GetState() == otelcol.StateClosed {
+				return nil
+			}
+
+			return errors.New("OTel Collector not in a closed state yet")
+		})
+
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to shutdown OTel Collector", "error", err, "state", oc.service.GetState())
+		} else {
+			slog.InfoContext(ctx, "OTel Collector shutdown", "state", oc.service.GetState())
+		}
 	}
-	<-oc.appDone
 
 	return nil
 }
 
 // Process an incoming Message Bus message in the plugin
-func (oc *Collector) Process(_ context.Context, msg *bus.Message) {
+func (oc *Collector) Process(ctx context.Context, msg *bus.Message) {
+	switch msg.Topic {
+	case bus.NginxConfigUpdateTopic:
+		oc.handleNginxConfigUpdate(ctx, msg)
+	default:
+		slog.DebugContext(ctx, "File plugin unknown topic", "topic", msg.Topic)
+	}
 }
 
 // Subscriptions returns the list of topics the plugin is subscribed to
 func (oc *Collector) Subscriptions() []string {
-	return []string{}
+	return []string{
+		bus.NginxConfigUpdateTopic,
+	}
+}
+
+func (oc *Collector) handleNginxConfigUpdate(ctx context.Context, msg *bus.Message) {
+	nginxConfigContext, ok := msg.Data.(*model.NginxConfigContext)
+	if !ok {
+		slog.ErrorContext(ctx, "Unable to cast message payload to *model.NginxConfigContext", "payload", msg.Data)
+	}
+
+	reloadCollector := oc.checkForNewNginxReceivers(nginxConfigContext)
+
+	if reloadCollector {
+		slog.InfoContext(ctx, "Reloading OTel collector config")
+		err := writeCollectorConfig(oc.config.Collector)
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to write OTel Collector config", "error", err)
+			return
+		}
+
+		oc.restartCollector(ctx)
+	}
+}
+
+func (oc *Collector) restartCollector(ctx context.Context) {
+	err := oc.Close(ctx)
+	if err != nil {
+		slog.ErrorContext(ctx, "Failed to shutdown OTel Collector", "error", err)
+		return
+	}
+
+	settings := OTelCollectorSettings(oc.config)
+	oTelCollector, err := otelcol.NewCollector(settings)
+	if err != nil {
+		slog.ErrorContext(ctx, "Failed to create OTel Collector", "error", err)
+		return
+	}
+	oc.service = oTelCollector
+
+	var runCtx context.Context
+	runCtx, oc.cancel = context.WithCancel(ctx)
+
+	go func() {
+		bootErr := oc.bootup(runCtx)
+		if bootErr != nil {
+			slog.ErrorContext(runCtx, "Unable to start OTel Collector", "error", bootErr)
+		}
+	}()
```

Comment:
Don't think we need to run `oc.bootup()` in a goroutine, as `bootup` starts the collector in its own goroutine.

Reply:
There is a for loop in the `oc.bootup()` function which will stop the function from exiting. Thats why we need to run `oc.bootup()` in a goroutine

Reply:
Yeah but that for loop only waits for a Running state and then returns a nil error. I would think we would block the `restartCollector` func in the same way that the `bootup` func does. Not a big deal if we keep this as-is though.

Reply:
oh never mind you are right the extra go routine is redundant. I missed `return nil` when the collector is in a running state